File: SinglyLinkedList/singlylinkedlist.py
"""
\Implement a singly linked list with the following operations:
(a) isEmpty(): Returns true if the list is empty, and false otherwise
(b) addToHead(data): Inserts an element to the beginning of the list
(c) addToTail(data): Inserts an element to the end of the list
(d) add(data, predecessor): Inserts an element after the given predecessor
node
(e) removeFromHead(): Removes the first node in the list
(f) remove(data): Removes the node with the given data
(g) retrieve(data, outputNodePointer): Returns the pointer to the node
with the requested data
(h) search(data): Returns true if the data exists in the list, and false
otherwise
(i) traverse(): Displays the contents of the list
Also, write a test program to check if the implementation works properly.
"""
class Node(object):

    def __init__(self,data):
    
        self.data = data;
        self.next = None;
class LinkedList(object):

    def __init__(self):

        self.head = None;

    def isEmpty(self): 

        if self.head == None:
            return True;
        else :
            return False;

    def addToHead(self,data):
        
        newNode = Node(data);
        if self.isEmpty():
            self.head=newNode;
        else:
            newNode.next = self.head;
            self.head = newNode;

    def addToTail(self,data):

        newNode = Node(data);
        if self.isEmpty():
            self.addToHead(data);
        else:
            tailNode = self.head;
            while tailNode.next != None:
                tailNode = tailNode.next
            tailNode.next=newNode;

    def add(self,data,predecessor):

        newNode = Node(data);
        newNode.next = predecessor.next;
        predecessor.next = newNode;

    def removeFromHead(self):

        if self.isEmpty():
            print("No Head to Remove")
        else:
            HeadNode  = self.head;        
            self.head =HeadNode.next ;
            del HeadNode.data;
            del HeadNode.next;

    def remove(self,data):

        if self.isEmpty():
            print("No Data to Remove")
        else:
            if(self.head.data==data):
                self.removeFromHead();
            else:
                currentNode =self.head;
                prevNode = None;
                if self.search(data):
                    while currentNode.data != data:
                        prevNode = currentNode;
                        currentNode = currentNode.next;
                    prevNode.next = currentNode.next
                    del currentNode.data;
                    del currentNode.next;
                else:
                    print("No Such Data to Remove")
            
    def retrieve(self,data):

        currentNode =self.head;
        while currentNode != None and currentNode.data != data:
            currentNode = currentNode.next;
        return currentNode;

    def search(self,data):

        counter =0;
        currentNode =self.head;
        while currentNode != None:
            if currentNode.data == data:
                counter +=1;
            currentNode = currentNode.next;
        if counter==0:
            return False;
        else:
            return True;

    def traverse(self):
        
        currentNode =self.head;
        while currentNode != None:
            print(currentNode.data);
            currentNode = currentNode.next;

File: SinglyLinkedList/c_implementation/singlylinkedlist.c
#include<stdio.h>
#include<stdlib.h>
#include "singlylinkedlist.h"

int isEmpty(LinkedList l_List){
    /* isEmpty(): Returns true if the list is empty, and false otherwise*/
    if(l_List.head == NULL){
        return 1;
    }
    return 0;
}

void addToHead(LinkedList *l_list,int data){
    /*Inserts an int data to the beginning of the list*/
    Node *newNode;
    newNode = malloc(sizeof(*newNode));
    newNode->data = data;
    newNode->next = NULL;

    if(isEmpty(*l_list)){
        l_list->head = newNode;
    }
    else{
        newNode->next =  l_list->head;
        l_list->head = newNode;
    }
}

void addToTail(LinkedList * l_list, int data){
    /*Inserts an element to the end of the list*/
    Node *newNode;
    newNode = malloc(sizeof(*newNode));
    newNode->data = data;
    newNode->next = NULL;

    if(isEmpty(*l_list)){
        l_list->head = malloc(sizeof(*newNode));
        l_list->head = newNode;
    }
    else{
        Node *tailNode = l_list->head;
        while(tailNode->next!=NULL){
           tailNode = tailNode->next; 
        }
        tailNode->next = newNode;
    }
}

void add(int data, Node *predecessor){
    /*Inserts an element after the given predecessor node*/
    Node *newNode;
    newNode = malloc(sizeof(*newNode));
    newNode->data = data;
    newNode->next = predecessor->next; 
    predecessor->next = newNode;
}

void removeFromHead(LinkedList *l_list){
    /*Removes the first node in the list*/
    if(isEmpty(*l_list)){
        printf("This List Is Empty\n");
    }
    else{
        Node *headNode = l_list->head;
        l_list->head = headNode->next;
        free(headNode);
    }
}

void removeData(LinkedList *l_list, int data){
    /* Removes the node with the given data*/

    if(isEmpty(*l_list)){
        printf("This List Is Empty\n");
    }
    else if(search(l_list,data)){
        if(l_list->head->data == data){
            removeFromHead(l_list);
        }
        else{
            Node *currentNode = l_list->head;
            Node *previousNode ;
            while(currentNode->data != data){
                previousNode = currentNode;
                currentNode = currentNode->next;
            }
            previousNode->next = currentNode->next;
            free(currentNode);
        }
    }
    else{
        printf("No Such Data to Remove\n");

    }
}

Node * retrieve(LinkedList *l_list, int data){
    /* Returns the pointer to the node with the requested data*/
    if(isEmpty(*l_list)){
        printf("This List Is Empty\n");
        return NULL;
    }
    else if(search(l_list,data)){
        Node *currentNode = l_list->head;
        while(currentNode->data != data){
            currentNode = currentNode->next;
        }
        return currentNode;
    }
    else{
        printf("No Such Data to Retrieve\n");
        return NULL;
    }
}

int search(LinkedList *l_list, int data){
    /*Returns true if the data exists in the list, and false otherwise*/
    if(isEmpty(*l_list)){
        printf("This List Is Empty\n");
        return 0;
    }
    else{
        Node *currentNode = l_list->head;
        while(currentNode->next !=  NULL){
            if(currentNode->data == data){
                return 1;
            }
            currentNode = currentNode->next;
        }
        if(currentNode->data == data){
            return 1;
        }
        return 0;
    }
}

void traverse(LinkedList l_list){
    /*Displays the contents of the list*/
    if(isEmpty(l_list)){
        printf("This List Is Empty\n");
    }
    else{
        Node *currentNode = l_list.head;
        while(currentNode->next != NULL){
            printf("%d, ",currentNode->data);
            currentNode = currentNode->next;
        }
        printf("%d \n",currentNode->data);
    }
}

void freeList(LinkedList *l_list){
    /*Free's the memory allocated to the list*/
    if(isEmpty(*l_list)){
        printf("This List Is Empty\n");
    }
    else{
        while(!isEmpty(*l_list)){
            removeFromHead(l_list);
        }
    }
}


